File: rust/src/compiler/parser.rs
use super::state::State;
use crate::scanner::token::TokenKind as TT;
use crate::scanner::Scanner;
use crate::value::*;
use std::collections::HashMap;
use std::io::Write;

#[derive(Default, Debug)]
pub struct ParseError {
    msg: String
}

impl ParseError {
    fn new(msg: &str) -> Self {
        Self { msg: msg.to_owned() }
    }

    pub fn get_message(&self) -> String {
        self.msg.clone()
    }
}


pub type Result = std::result::Result<(), ParseError>;

pub struct Parser<'a, W: Write> {
    pub(super) scanner: Scanner<'a>,
    pub(super) out: &'a mut W,
    pub(super) state: State,
    pub(crate) variables: HashMap<String, String>,
}

impl<'a, W: Write> Parser<'a, W> {
    pub fn new(
        source: &'a str,
        out: &'a mut W,
        variables: HashMap<String, String>,
    ) -> Parser<'a, W> {
        Parser {
            scanner: Scanner::new(source),
            out,
            state: State::new(),
            variables,
        }
    }

    pub fn parse_to_out(&mut self) -> Result {
        let _ = write!(self.out, "{}",  RESET);
        loop {
            let token = self.scanner.scan_token();
            match token.kind {
                TT::Text => {
                    let _ = write!(self.out, "{}", token.content);
                }
                TT::B => {
                    self.state.push(token.content, B);
                    let _ = write!(self.out, "{}", B);

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::I => {
                    self.state.push(token.content, I);
                    let _ = write!(self.out, "{}", I);

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::S => {
                    self.state.push(token.content, S);
                    let _ = write!(self.out, "{}", S);

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::U => {
                    self.state.push(token.content, U);
                    let _ = write!(self.out, "{}", U);

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::C => {
                    let token = self.scanner.scan_token();
                    match token.kind {
                        TT::Dot => {
                            let token = self.scanner.scan_token();
                            match token.kind {
                                TT::Black => {
                                    self.state.push("c", C::BLACK);
                                    let _ = write!(self.out, "{}", C::BLACK);
                                }
                                TT::Blue => {
                                    self.state.push("c", C::BLUE);
                                    let _ = write!(self.out, "{}", C::BLUE);
                                }
                                TT::Cyan => {
                                    self.state.push("c", C::CYAN);
                                    let _ = write!(self.out, "{}", C::CYAN);
                                }
                                TT::Green => {
                                    self.state.push("c", C::GREEN);
                                    let _ = write!(self.out, "{}", C::GREEN);
                                }
                                TT::Magenta => {
                                    self.state.push("c", C::MAGENTA);
                                    let _ = write!(self.out, "{}", C::MAGENTA);
                                }
                                TT::Red => {
                                    self.state.push("c", C::RED);
                                    let _ = write!(self.out, "{}", C::RED);
                                }
                                TT::White => {
                                    self.state.push("c", C::WHITE);
                                    let _ = write!(self.out, "{}", C::WHITE);
                                }
                                TT::Yellow => {
                                    self.state.push("c", C::YELLOW);
                                    let _ = write!(self.out, "{}", C::YELLOW);
                                }
                                TT::Rgb => {
                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::LeftParen {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let r: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Comma {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let g: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Comma {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let b: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::RightParen {
                                        return Err(ParseError::default())
                                    }

                                    let rgb = C::rgb(r, g, b);
                                    self.state.push("c", &rgb);
                                    let _ = write!(self.out, "{}", &rgb);
                                }
                                _ => return Err(ParseError::default()),
                            }
                        }
                        _ => return Err(ParseError::default()),
                    }

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::X => {
                    let token = self.scanner.scan_token();
                    match token.kind {
                        TT::Dot => {
                            let token = self.scanner.scan_token();
                            match token.kind {
                                TT::Black => {
                                    self.state.push("x", X::BLACK);
                                    let _ = write!(self.out, "{}", X::BLACK);
                                }
                                TT::Blue => {
                                    self.state.push("x", X::BLUE);
                                    let _ = write!(self.out, "{}", X::BLUE);
                                }
                                TT::Cyan => {
                                    self.state.push("x", X::CYAN);
                                    let _ = write!(self.out, "{}", X::CYAN);
                                }
                                TT::Green => {
                                    self.state.push("x", X::GREEN);
                                    let _ = write!(self.out, "{}", X::GREEN);
                                }
                                TT::Magenta => {
                                    self.state.push("x", X::MAGENTA);
                                    let _ = write!(self.out, "{}", X::MAGENTA);
                                }
                                TT::Red => {
                                    self.state.push("x", X::RED);
                                    let _ = write!(self.out, "{}", X::RED);
                                }
                                TT::White => {
                                    self.state.push("x", X::WHITE);
                                    let _ = write!(self.out, "{}", X::WHITE);
                                }
                                TT::Yellow => {
                                    self.state.push("x", X::YELLOW);
                                    let _ = write!(self.out, "{}", X::YELLOW);
                                }
                                TT::Rgb => {
                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::LeftParen {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let r: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Comma {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let g: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Comma {
                                        return Err(ParseError::default())
                                    }

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::Number {
                                        return Err(ParseError::default())
                                    }
                                    let b: u8 = token.content.parse().unwrap();

                                    let token = self.scanner.scan_token();
                                    if token.kind != TT::RightParen {
                                        return Err(ParseError::default())
                                    }

                                    let rgb = X::rgb(r, g, b);
                                    self.state.push("x", &rgb);
                                    let _ = write!(self.out, "{}", &rgb);
                                }
                                _ => return Err(ParseError::default()),
                            }
                        }
                        _ => return Err(ParseError::default()),
                    }

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::Slash => {
                    let token = self.scanner.scan_token();
                    match token.kind {
                        TT::B => {
                            if self.state.current_tag() != "b" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let _ = write!(self.out, "{}", RESET_B);
                        }
                        TT::I => {
                            if self.state.current_tag() != "i" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let _ = write!(self.out, "{}", RESET_I);
                        }
                        TT::S => {
                            if self.state.current_tag() != "s" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let _ = write!(self.out, "{}", RESET_S);
                        }
                        TT::U => {
                            if self.state.current_tag() != "u" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let _ = write!(self.out, "{}", RESET_U);
                        }
                        TT::C => {
                            if self.state.current_tag() != "c" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let saved = self.state.current_save();
                            let _ = write!(self.out, "{}", saved);
                        }
                        TT::X => {
                            if self.state.current_tag() != "x" {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let saved = self.state.current_save();
                            let _ = write!(self.out, "{}", saved);
                        }
                        TT::Identifier
                        | TT::Black
                        | TT::Blue
                        | TT::Cyan
                        | TT::Green
                        | TT::Magenta
                        | TT::Red
                        | TT::Rgb
                        | TT::White
                        | TT::Yellow => {
                            if self.state.current_tag() != token.content {
                                return Err(ParseError::default())
                            }
                            self.state.pop();
                            let saved = self.state.current_save();
                            let _ = write!(self.out, "{}", saved);
                        }
                        _ => return Err(ParseError::default()),
                    }

                    let token = self.scanner.scan_token();
                    if token.kind != TT::CloseTag {
                        return Err(ParseError::default())
                    }
                }
                TT::Identifier
                | TT::Black
                | TT::Blue
                | TT::Cyan
                | TT::Green
                | TT::Magenta
                | TT::Red
                | TT::Rgb
                | TT::White
                | TT::Yellow => {
                    let var = self.variables.get(token.content);
                    if let Some(val) = var {
                        self.state.push(token.content, val);
                        let _ = write!(self.out, "{}", val);
                    } else {
                        return Err(ParseError::default())
                    }
                }
                TT::Error => {
                    return Err(ParseError { msg: format!("Unexpected character. {}", token.content) });
                }
                TT::Eof => {
                    let _ = write!(self.out, "{}", RESET);
                    return Ok(());
                }
                _ => continue,
            }
        }
    }
}

File: javascript/src/compiler/parser.ts
import { State } from "./state";
import { TokenKind as TT } from "../scanner/token";
import { Scanner } from "../scanner";
import {
    B,
    C,
    I,
    RESET,
    RESET_B,
    RESET_I,
    RESET_S,
    RESET_U,
    S,
    U,
    X,
} from "../value";

class ParseError extends Error {}

export class Parser {
    scanner: Scanner;
    out: NodeJS.WriteStream;
    state: State = new State();
    variables: Record<string, string>;

    constructor(
        source: string,
        out: NodeJS.WriteStream,
        variables: Record<string, string>
    ) {
        this.scanner = new Scanner(source);
        this.out = out;
        this.variables = variables;
    }

    parseToOut() {
        this.out.write(RESET);
        for (;;) {
            let token = this.scanner.scanToken();
            switch (token.kind) {
                case TT.Text:
                    this.out.write(token.content);
                    break;
                case TT.B:
                    this.state.push(token.content, B);
                    this.out.write(B);

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.I: {
                    this.state.push(token.content, I);
                    this.out.write(I);

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                }
                case TT.S:
                    this.state.push(token.content, S);
                    this.out.write(S);

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.U:
                    this.state.push(token.content, U);
                    this.out.write(U);

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.C:
                    token = this.scanner.scanToken();
                    switch (token.kind) {
                        case TT.Dot:
                            token = this.scanner.scanToken();
                            switch (token.kind) {
                                case TT.Black:
                                    this.state.push("c", C.BLACK);
                                    this.out.write(C.BLACK);
                                    break;
                                case TT.Blue:
                                    this.state.push("c", C.BLUE);
                                    this.out.write(C.BLUE);
                                    break;
                                case TT.Cyan:
                                    this.state.push("c", C.CYAN);
                                    this.out.write(C.CYAN);
                                    break;
                                case TT.Green:
                                    this.state.push("c", C.GREEN);
                                    this.out.write(C.GREEN);
                                    break;
                                case TT.Magenta:
                                    this.state.push("c", C.MAGENTA);
                                    this.out.write(C.MAGENTA);
                                    break;
                                case TT.Red:
                                    this.state.push("c", C.RED);
                                    this.out.write(C.RED);
                                    break;
                                case TT.White:
                                    this.state.push("c", C.WHITE);
                                    this.out.write(C.WHITE);
                                    break;
                                case TT.Yellow:
                                    this.state.push("c", C.YELLOW);
                                    this.out.write(C.YELLOW);
                                    break;
                                case TT.Rgb:
                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.LeftParen)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const r: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Comma)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const g: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Comma)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const b: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.RightParen)
                                        throw new ParseError();

                                    const rgb = C.rgb(r, g, b);
                                    this.state.push("c", rgb);
                                    this.out.write(rgb);
                                    break;
                                default:
                                    throw new ParseError();
                            }
                            break;
                        default:
                            throw new ParseError();
                    }
                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;

                case TT.X:
                    token = this.scanner.scanToken();
                    switch (token.kind) {
                        case TT.Dot:
                            token = this.scanner.scanToken();
                            switch (token.kind) {
                                case TT.Black:
                                    this.state.push("x", X.BLACK);
                                    this.out.write(X.BLACK);
                                    break;
                                case TT.Blue:
                                    this.state.push("x", X.BLUE);
                                    this.out.write(X.BLUE);
                                    break;
                                case TT.Cyan:
                                    this.state.push("x", X.CYAN);
                                    this.out.write(X.CYAN);
                                    break;
                                case TT.Green:
                                    this.state.push("x", X.GREEN);
                                    this.out.write(X.GREEN);
                                    break;
                                case TT.Magenta:
                                    this.state.push("x", X.MAGENTA);
                                    this.out.write(X.MAGENTA);
                                    break;
                                case TT.Red:
                                    this.state.push("x", X.RED);
                                    this.out.write(X.RED);
                                    break;
                                case TT.White:
                                    this.state.push("x", X.WHITE);
                                    this.out.write(X.WHITE);
                                    break;
                                case TT.Yellow:
                                    this.state.push("x", X.YELLOW);
                                    this.out.write(X.YELLOW);
                                    break;
                                case TT.Rgb:
                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.LeftParen)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const r: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Comma)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const g: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Comma)
                                        throw new ParseError();

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.Number)
                                        throw new ParseError();
                                    const b: number = parseInt(token.content);

                                    token = this.scanner.scanToken();
                                    if (token.kind != TT.RightParen)
                                        throw new ParseError();

                                    const rgb = X.rgb(r, g, b);
                                    this.state.push("x", rgb);
                                    this.out.write(rgb);
                                    break;
                                default:
                                    throw new ParseError();
                            }
                            break;
                        default:
                            throw new ParseError();
                    }
                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.Slash:
                    token = this.scanner.scanToken();
                    switch (token.kind) {
                        case TT.B:
                            if (this.state.currentTag() != "b")
                                throw new ParseError();
                            this.state.pop();
                            this.out.write(RESET_B);
                            break;
                        case TT.I:
                            if (this.state.currentTag() != "i")
                                throw new ParseError();
                            this.state.pop();
                            this.out.write(RESET_I);
                            break;
                        case TT.S:
                            if (this.state.currentTag() != "s")
                                throw new ParseError();
                            this.state.pop();
                            this.out.write(RESET_S);
                            break;
                        case TT.U:
                            if (this.state.currentTag() != "u")
                                throw new ParseError();
                            this.state.pop();
                            this.out.write(RESET_U);
                            break;
                        case TT.C: {
                            if (this.state.currentTag() != "c")
                                throw new ParseError();
                            this.state.pop();
                            let saved = this.state.currentSave();
                            this.out.write(saved);
                            break;
                        }
                        case TT.X: {
                            if (this.state.currentTag() != "x")
                                throw new ParseError();
                            this.state.pop();
                            let saved = this.state.currentSave();
                            this.out.write(saved);
                            break;
                        }
                        case TT.Identifier:
                        case TT.Black:
                        case TT.Blue:
                        case TT.Cyan:
                        case TT.Green:
                        case TT.Magenta:
                        case TT.Red:
                        case TT.Rgb:
                        case TT.White:
                        case TT.Yellow: {
                            if (this.state.currentTag() != token.content)
                                throw new ParseError();
                            this.state.pop();
                            let saved = this.state.currentSave();
                            this.out.write(saved);
                            break;
                        }
                        default:
                            throw new ParseError();
                    }

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.Identifier:
                case TT.Black:
                case TT.Blue:
                case TT.Cyan:
                case TT.Green:
                case TT.Magenta:
                case TT.Red:
                case TT.Rgb:
                case TT.White:
                case TT.Yellow:
                    const val = this.variables[token.content];
                    if (typeof val != "undefined") {
                        this.state.push(token.content, val);
                        this.out.write(val);
                    } else {
                        throw new ParseError();
                    }

                    token = this.scanner.scanToken();
                    if (token.kind != TT.CloseTag) throw new ParseError();
                    break;
                case TT.Error:
                    throw new ParseError("Unexpected character. " + token.content)
                case TT.Eof:
                    this.out.write(RESET);
                    return;
                default:
                    continue;
            }
        }
    }
}
